File: sdk/python/onecortex/_http.py
import time
import httpx
from .exceptions import (
    OnecortexError, NotFoundError, AlreadyExistsError,
    InvalidArgumentError, UnauthorizedError, PermissionDeniedError, OnecortexServerError
)

_ERROR_MAP = {
    "NOT_FOUND": NotFoundError,
    "ALREADY_EXISTS": AlreadyExistsError,
    "INVALID_ARGUMENT": InvalidArgumentError,
    "UNAUTHENTICATED": UnauthorizedError,
    "PERMISSION_DENIED": PermissionDeniedError,
}

def _raise_for_response(response: httpx.Response) -> None:
    if response.status_code < 400:
        return
    try:
        body = response.json()
        code = body.get("error", {}).get("code", "UNKNOWN")
        message = body.get("error", {}).get("message", response.text)
    except Exception:
        code = "UNKNOWN"
        message = response.text

    exc_class = _ERROR_MAP.get(code, OnecortexServerError)
    raise exc_class(message, status_code=response.status_code)


class HttpClient:
    """Synchronous httpx client with retry logic and auth header injection."""

    def __init__(self, api_key: str, host: str):
        self._host = host.rstrip("/")
        self._headers = {
            "Api-Key": api_key,
            "Content-Type": "application/json",
        }
        self._client = httpx.Client(headers=self._headers, timeout=30.0)

    def request(self, method: str, path: str, **kwargs) -> httpx.Response:
        url = f"{self._host}{path}"
        # Retry on 429 and 5xx with exponential backoff: 1s, 2s, 4s
        delays = [1, 2, 4]
        last_exc = None
        for attempt, delay in enumerate([0] + delays):
            if delay:
                time.sleep(delay)
            try:
                response = self._client.request(method, url, **kwargs)
                if response.status_code in (429,) or response.status_code >= 500:
                    if attempt < len(delays):
                        last_exc = response
                        continue
                _raise_for_response(response)
                return response
            except (httpx.ConnectError, httpx.TimeoutException) as e:
                last_exc = e
                continue
        if isinstance(last_exc, httpx.Response):
            _raise_for_response(last_exc)
        raise OnecortexServerError(f"Request failed after retries: {last_exc}")

    def get(self, path: str, **kwargs) -> httpx.Response:
        return self.request("GET", path, **kwargs)

    def post(self, path: str, json: dict | None = None, **kwargs) -> httpx.Response:
        return self.request("POST", path, json=json, **kwargs)

    def delete(self, path: str, **kwargs) -> httpx.Response:
        return self.request("DELETE", path, **kwargs)

    def patch(self, path: str, json: dict | None = None, **kwargs) -> httpx.Response:
        return self.request("PATCH", path, json=json, **kwargs)

File: sdk/typescript/src/http.ts
import type { OnecortexConfig } from './types';

export class OnecortexHttpError extends Error {
  constructor(
    public readonly statusCode: number,
    public readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = 'OnecortexHttpError';
  }
}

const RETRY_DELAYS = [1000, 2000, 4000];

export class HttpClient {
  private readonly host: string;
  private readonly headers: Record<string, string>;

  constructor(config: OnecortexConfig) {
    this.host = (config.host ?? 'http://localhost:8080').replace(/\/$/, '');
    this.headers = {
      'Api-Key': config.apiKey,
      'Content-Type': 'application/json',
    };
  }

  private async doFetch(method: string, path: string, body?: unknown): Promise<unknown> {
    const url = `${this.host}${path}`;
    const init: RequestInit = {
      method,
      headers: this.headers,
      body: body !== undefined ? JSON.stringify(body) : undefined,
    };

    for (let attempt = 0; attempt <= RETRY_DELAYS.length; attempt++) {
      if (attempt > 0) {
        await new Promise((r) => setTimeout(r, RETRY_DELAYS[attempt - 1]));
      }

      const res = await fetch(url, init);

      if (res.status === 429 || res.status >= 500) {
        if (attempt < RETRY_DELAYS.length) continue;
      }

      if (!res.ok) {
        let code = 'UNKNOWN';
        let message = res.statusText;
        try {
          const json = (await res.json()) as { error?: { code?: string; message?: string } };
          code = json.error?.code ?? code;
          message = json.error?.message ?? message;
        } catch {}
        throw new OnecortexHttpError(res.status, code, message);
      }

      // 202 Accepted and 204 No Content may have no body
      if (res.status === 202 || res.status === 204) return {};

      const text = await res.text();
      return text ? JSON.parse(text) : {};
    }

    throw new OnecortexHttpError(500, 'INTERNAL', 'Request failed after retries');
  }

  async get(path: string, params?: Record<string, string>): Promise<unknown> {
    const url = params ? `${path}?${new URLSearchParams(params)}` : path;
    return this.doFetch('GET', url);
  }

  async post(path: string, body?: unknown): Promise<unknown> {
    return this.doFetch('POST', path, body);
  }

  async delete(path: string): Promise<unknown> {
    return this.doFetch('DELETE', path);
  }

  async patch(path: string, body?: unknown): Promise<unknown> {
    return this.doFetch('PATCH', path, body);
  }
}
